lib/plugins/aws/invokeLocal/index.js

```diff
@@ -267,62 +259,123 @@ class AwsInvokeLocal {
-              );
-          }
-          return downloadPromise
-            .then(() => fse.copySync(layerContentsCachePath, layerContentsPath))
+
+          return dirExists(layerContentsPath)
+            .then(exists => {
+              if (exists) {
+                return layerContentsPath;
+              }
+              return dirExists(layerContentsCachePath).then(cacheExists => {
+                if (!cacheExists) {
+                  this.serverless.cli.log(`Downloading layer ${layer}...`);
+                  return mkdirpAsync(path.join(layerContentsCachePath))
+                    .then(() =>
+                      this.provider.request('Lambda', 'getLayerVersion', {
+                        LayerName: layerArn,
+                        VersionNumber: layerVersion,
+                      })
+                    )
+                    .then(layerInfo =>
+                      download(layerInfo.Content.Location, layerContentsCachePath, {
+                        extract: true,
+                      })
+                    );
+                }
+                return null;
+              });
+            })
+            .then(() => fse.copy(layerContentsCachePath, layerContentsPath))
             .then(() => layerContentsPath);
         }
       )
     );
   }
 
+  getDockerImageName() {
+    return `sls-docker-${this.getRuntime()}`;
+  }
+
   buildDockerImage(layerPaths) {
     const runtime = this.getRuntime();
-    const imageName = 'sls-docker';
+    const imageName = this.getDockerImageName();
 
-    return new BbPromise((resolve, reject) => {
-      let dockerfile = `FROM lambci/lambda:${runtime}`;
-      for (const layerPath of layerPaths) {
-        dockerfile += `\nADD --chown=sbx_user1051:495 ${layerPath} /opt`;
-      }
-      mkdirp.sync(path.join('.serverless', 'invokeLocal'));
-      const dockerfilePath = path.join('.serverless', 'invokeLocal', 'Dockerfile');
-      fs.writeFileSync(dockerfilePath, dockerfile);
-      this.serverless.cli.log('Building Docker image...');
-      const docker = spawn('docker', [
-        'build',
-        '-t',
-        imageName,
-        `${this.serverless.config.servicePath}`,
-        '-f',
-        dockerfilePath,
-      ]);
-      docker.on('exit', error => {
-        return error ? reject(error) : resolve(imageName);
+    let dockerFileContent = `FROM lambci/lambda:${runtime}`;
+    for (const layerPath of layerPaths) {
+      dockerFileContent += `\nADD --chown=sbx_user1051:495 ${layerPath} /opt`;
+    }
+
+    const dockerfilePath = path.join('.serverless', 'invokeLocal', runtime, 'Dockerfile');
+    const dockerfileCachePath = path.join(cachePath, 'dockerfiles', runtime, 'Dockerfile');
+
+    return fileExists(dockerfileCachePath)
+      .then(exists => {
+        if (exists) {
+          return fs.readFileAsync(dockerfileCachePath).then(contents => {
+            return contents.toString() === dockerFileContent && this.checkDockerImage(imageName);
```

Comment:
Let's put this conditional in more literal form (to avoid confusion) as:

```javascript
if (contents.toString() === dockerFileContent) return this.checkDockerImage(imageName);
```